Source/Core/DolphinQt2/Resources.cpp

```diff
@@ -2,46 +2,74 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include <QGuiApplication>
+#include <QIcon>
+#include <QScreen>
 #include <QStringList>
 
 #include "Common/CommonPaths.h"
 #include "Common/FileUtil.h"
 #include "DolphinQt2/Resources.h"
+#include "DolphinQt2/Settings.h"
 
 QList<QPixmap> Resources::m_platforms;
 QList<QPixmap> Resources::m_countries;
 QList<QPixmap> Resources::m_ratings;
 QList<QPixmap> Resources::m_misc;
 
+QString Resources::GetPreferredPixmapScale(int max)
+{
+  const auto dpr = QGuiApplication::primaryScreen()->devicePixelRatio();
+  const auto scale = std::min(dpr <= 1 ? 1 : (dpr <= 2) ? 2 : 4, max);
+
+  if (scale > 1)
+    return QStringLiteral("@%1x").arg(scale);
+  else
+    return QStringLiteral("");
+}
+
+QPixmap Resources::GetScaledPixmap(const std::string& name, bool is_themed)
+{
+  QPixmap pixmap;
+  for (int scale : {4, 2, 1})
+  {
+    pixmap = QPixmap(QStringLiteral("%1%2%3.png")
```

Comment:
Pixmaps at @4x do not have their devicePixelRatio automatically set by Qt, and need to be set manually.